aspnetcore/web-api/handle-errors.md

```diff
@@ -140,13 +142,101 @@ For web API controllers, MVC responds with a <xref:Microsoft.AspNetCore.Mvc.Vali
+}
+```
+
+The next section shows how to customize the problem details response body to return a more helpful response.
+
+<a name="cpd7"></a>
+
+#### Customize problem details with `CustomizeProblemDetails`
+
+The following code uses <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions> to set <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails>:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_api_controller" highlight="7-29":::
+
+The updated API controller:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Controllers/ValuesController.cs" id="snippet"  highlight="9-17,27-35":::
+
+A problem details response is generated with the previous code when any of the following conditions apply:
+
+* The `/api/values/divide` or the `/divide` endpoint is called with a zero denominator.
+* The `/api/values/squareroot` or the `/squareroot` endpoint is called with a radicand less than zero.
+* The URI has no matching endpoint.
+
+The problem details response body contains the following when either `squareroot` endpoint is called with a radicand less than zero:
+
+```json
+{
+  "type": "https://en.wikipedia.org/wiki/Square_root",
+  "title": "Bad Input",
+  "status": 400,
+  "detail": "Negative or complex numbers are not allowed."
+}
+```
+
+#### Generate problem details from Middleware
+
+An alternative approach to using <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions> to set <xref:Microsoft.AspNetCore.Http.ProblemDetailsOptions.CustomizeProblemDetails> is to set the <xref:Microsoft.AspNetCore.Http.ProblemDetailsContext.ProblemDetails> in middleware. A problem details response can be written by calling [`IProblemDetailsService.WriteAsync`](/dotnet/api/microsoft.aspnetcore.http.iproblemdetailsservice.writeasync?view=aspnetcore-7.0&preserve-view=true):
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/problem-details-service/Program.cs" id="snippet_middleware" highlight="6,19-48":::
+
+In the preceding code, the minimal API endpoints `/divide` and `/squareroot` return the expected custom problem response on error input. The API controller endpoints return the default problem response on error input, not the custom problem response. The default problem response is returned because the API controller has written to the response stream before [`IProblemDetailsService.WriteAsync`](https://github.com/dotnet/aspnetcore/blob/ce2db7ea0b161fc5eb35710fca6feeafeeac37bc/src/Http/Http.Extensions/src/ProblemDetailsService.cs#L24) is called. The [`IProblemDetailsService.WriteAsync` source code](https://github.com/dotnet/aspnetcore/blob/ce2db7ea0b161fc5eb35710fca6feeafeeac37bc/src/Http/Http.Extensions/src/ProblemDetailsService.cs#L24) shows that if the HTTP response has started, the custom problem details are not written.
```

Comment:
```suggestion
In the preceding code, the minimal API endpoints `/divide` and `/squareroot` return the expected custom problem response on error input. 

The API controller endpoints return the default problem response on error input, not the custom problem response. The default problem response is returned because the API controller has written to the response stream, [Problem details for error status codes](https://learn.microsoft.com/en-us/aspnet/core/web-api/?view=aspnetcore-7.0#problem-details-for-error-status-codes-1), before [`IProblemDetailsService.WriteAsync`](https://github.com/dotnet/aspnetcore/blob/ce2db7ea0b161fc5eb35710fca6feeafeeac37bc/src/Http/Http.Extensions/src/ProblemDetailsService.cs#L24) is called and the response will not be written again.
```